### PR TITLE
[#36] Warn user on possible timeshifts

### PR DESCRIPTION
**Pull request by YuriRomanowski**
## Description
Problem: In some cases the bot selects the date rather arbitrarily, and if a timeshift is somewhere around, the translation may be dependent on concrete date. We should warn a user about possible timeshifts.

Solution: For some types of date references, also compute a date range to check for the possible timeshift, and report to user if something found.

## Related issue(s)

Fixed part of #36

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: src/TzBot/TZ.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      desiredTimeshifts =
+        takeWhile (\(_ix, t) -> t <= after) $
+          zip [ixb..] $ VU.toList startingFromIxb
+      timeZones = fromJust $ nonEmpty $
+        flip map desiredTimeshifts (\(ix, int64utc) ->
+          ( int64ToUtcTime int64utc
+          , Offset $ timeZoneMinutes $ timeZoneForIx tz ix)
+          )
+      timeShifts = zip (toList timeZones) $ tail timeZones
+  map (\((_ub, b), (ua, a)) -> TimeShift ua b a) timeShifts
```
Comment (YuriRomanowski) [inline src/TzBot/TZ.hs:45]: If this function suites our need, we can then create an issue [in the original repo](https://github.com/ysangkok/haskell-tz/issues?q=is%3Aissue+is%3Aclosed) to add it there.

File: test/Test/TzBot/RenderSpec.hs [hunk 2]
```diff
@@ -71,11 +74,38 @@ test_renderSpec = TestGroup "Render" $
...
         "02:00, Saturday, 21 January 2023 in America/Havana"
     ]
+
+  , TestGroup "TimeshiftWarning"
+    [ testCase "Implicit date" $
+      mkTestCase nearTimeshift "10am" userMoscow userHavana
+      [ translWithCommonNote
+          "\"10am\", 11 March 2023 in Europe/Moscow"
+          "02:00, Saturday, 11 March 2023 in America/Havana"
+          "_Warning! There is a time shift in America/Havana at 2023-03-12 05:00:00 UTC, so around this time conversion may depend on concrete date._"
```
Comment (dcastro) [inline test/Test/TzBot/RenderSpec.hs:84]: > There is a time shift in America/Havana at 2023-03-12 05:00:00 UTC

This doesn't seem to be correct. According to [this](https://www.timeanddate.com/time/zone/cuba/havana), the clocks are turned forward 1 hour from midnight to 1am.

**Comment by dcastro** [comment]
I think I missed the conversation that led to this PR. Can you please point me to the context? 

> In some cases the bot selects the date rather arbitrarily, and if a timeshift is somewhere around, the translation may be dependent on concrete date. We should warn a user about possible timeshifts.

Also, you mentioned "in some cases", can you please give me a concrete example to see when/why this is useful? And how would you define "somewhere around"?

Sorry, I just think I'm missing some context to understand what this does and why :sweat_smile:

File: test/Test/TzBot/RenderSpec.hs (hunk 2 above)
Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:84]: Yes, I wrote "UTC", which means +5 hours relative to Havana. Did you take this into account in your comment?

Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:84]: Actually, I must admit that using UTC for this maybe not so convenient for the end user, but it's the reliable thing for defining timeshift time. But I think we could just convert the timeshift UTCTime to the receiver's timezone, and show it instead. Does it make sense?

File: test/Test/TzBot/RenderSpec.hs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+          "03:00, Sunday, 12 March 2023 in America/Havana"
+          "_Warning! There is a time shift in America/Havana at 2023-03-12 05:00:00 UTC, so around this time conversion may depend on concrete date._"
+      ]
+    , testCase "Day of week, inversed users" $
+      mkTestCase nearTimeshift "10am on sunday" userHavana userMoscow
+      [ translWithCommonNote
+          "\"10am on sunday\", 12 March 2023 in America/Havana"
+          "17:00, Sunday, 12 March 2023 in Europe/Moscow"
+          "_Warning! There is a time shift in America/Havana at 2023-03-12 05:00:00 UTC, so around this time conversion may depend on concrete date._"
+      ]
```
Comment (dcastro) [inline test/Test/TzBot/RenderSpec.hs:99]: We should also add a test for when there is a time change in _both_ the source and the target time zones, e.g.:

```
    , testCase "Demo" $
      mkTestCase 
        (toUTC [tz|2023-03-25 15:00:00 [Europe/London]|])
        "10am"
        (User {uId="london", uIsBot=False, uTz=Europe__London})
        (User {uId="lisbon", uIsBot=False, uTz=Europe__Lisbon})
      [ translWithCommonNote
          "\"10am\", 26 March 2023 in Europe/London"
          "10:00, Sunday, 26 March 2023 in Europe/Lisbon"
          [int|s|
            _Warning! There is a time shift in Europe/London at 2023-03-26 01:00:00 UTC, so around this time conversion may depend on concrete date._
            _Warning! There is a time shift in Europe/Lisbon at 2023-03-26 01:00:00 UTC, so around this time conversion may depend on concrete date._
          |]
      ]
```

File: test/Test/TzBot/RenderSpec.hs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        "02:00, Saturday, 21 January 2023 in America/Havana"
+    ]
+
+  , TestGroup "TimeshiftWarning"
+    [ testCase "Implicit date" $
+      mkTestCase nearTimeshift "10am" userMoscow userHavana
+      [ translWithCommonNote
+          "\"10am\", 11 March 2023 in Europe/Moscow"
+          "02:00, Saturday, 11 March 2023 in America/Havana"
+          "_Warning! There is a time shift in America/Havana at 2023-03-12 05:00:00 UTC, so around this time conversion may depend on concrete date._"
```
Comment (dcastro) [inline test/Test/TzBot/RenderSpec.hs:84]: From a user's point of view, it may not be clear what this means, what the implications are, or what the user should do about it:

> so around this time conversion may depend on concrete date

I think we should be very explicit about what the warning means and why we're warning the user about this time change. What do you think of this more verbose explanation?

> Warning: We inferred "10am" refers to 11 March 2023 in Europe/Moscow and converted it to America/Havana,
> but there is a time change near this date:
>
>   * at 01:00, 12 March 2023 in Europe/Moscow, the clocks are turned forward 1 hour(s).
>
> Beware that if this inferrence is not correct and the sender meant a different date, the conversion may not be accurate.


Now we're being very clear that we're emitting a warning _because_ we inferred the date, and that, in this specific case, the conversion is sensitive to this inference - if we inferred the wrong date, the conversion is likely to be wrong.

Note that we'd use 2 bullet points if there is a time change in _both_ the source and the target time zones, for example (I made up data for this example):

> Warning: We inferred "10am" refers to 11 March 2023 in Europe/Moscow and converted it to America/Havana,
> but there is a time change near this date:
>
>   * at 01:00, 12 March 2023 in Europe/Moscow, the clocks are turned forward 1 hour(s).
>   * at 02:00, 12 March 2023 in America/Havana, the clocks are turned forward 2 hour(s).
>
> Beware that if this inferrence is not correct and the sender meant a different date, the conversion may not be accurate.

File: test/Test/TzBot/RenderSpec.hs (hunk 3 above)
Comment (dcastro) [reply test/Test/TzBot/RenderSpec.hs:99]: Let's also add tests that check that when the sender explicitly says "tomorrow", or "in 3 days", or explicitly specifies a date like "26 March", then the note is **not** emitted.

File: test/Test/TzBot/RenderSpec.hs (hunk 2 above)
Comment (dcastro) [reply test/Test/TzBot/RenderSpec.hs:84]: Oh, I missed the "UTC" bit at the end :sweat_smile: 

> Actually, I must admit that using UTC for this is maybe not so convenient for the end user, but it's the reliable thing for defining timeshift time. But I think we could just convert the timeshift UTCTime to the receiver's timezone, and show it instead. Does it make sense?

I agree using UTC here is really inconvenient for users.

I think we should show the time in the timezone where the shift occurred. See [my other comment below](https://github.com/serokell/tzbot/pull/58#discussion_r1115529130).

File: test/Test/TzBot/RenderSpec.hs (hunk 4 above)
Comment (dcastro) [inline test/Test/TzBot/RenderSpec.hs:84]: I also wanted to point out the inconsistent formats we're using here:

>           "02:00, Saturday, 11 March 2023 in America/Havana"
>          "_Warning! There is a time shift in America/Havana at 2023-03-12 05:00:00 UTC, so around this time conversion may depend on concrete date._"

In the first line, we're using `HH:MM, day of week, DD Month YY` and in the second we're using `YYYY-MM-DD HH:MM:SS`. Let's try to always use the first format for consistency.

Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:84]: Great remark! I agree, let's elaborate to make more clear for user what we really want to say to him :)

Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:134]: It looks like this:
изображение

Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:134]: Now I use DD Month YY since I think time and day of week are not required.

File: test/Test/TzBot/RenderSpec.hs [hunk 5]
```diff
@@ -119,6 +123,73 @@ test_renderSpec = TestGroup "Render"
...
       ]
     ]
+
+  , TestGroup "TimeshiftWarning"
+    [ testCase "Implicit date" $
+      mkChatCase nearTimeshift "10am" userMoscow userHavana
+      [ translWithCommonNote
+          "\"10am\", 11 March 2023 in Europe/Moscow"
+          "02:00, Saturday, 11 March 2023 in America/Havana"
+          "_Warning: We inferred that \"10am\" refers to 11 March 2023 in Europe/Moscow and converted it to America/Havana, but there is a time change near this date_:\n  \8226 _At 01:00, 12 March 2023 in America/Havana, the clocks are turned forward 1 hour(s)_.\n_Beware that if this inference is not correct and the sender meant a different date, the conversion may not be accurate._"
```
Comment (dcastro) [inline test/Test/TzBot/RenderSpec.hs:133]: Please use an interpolator and replace `\8226` with `•` to make this easier to read (in this test and the other ones where we use more complex multi-line strings)

```hs
          [int|s|  
            _Warning: We inferred that "10am" refers to 11 March 2023 in Europe/Moscow and converted it to America/Havana, but there is a time change near this date_:
              • _At 01:00, 12 March 2023 in America/Havana, the clocks are turned forward 1 hour(s)_.
            _Beware that if this inference is not correct and the sender meant a different date, the conversion may not be accurate._
          |]
```

Comment (dcastro) [inline test/Test/TzBot/RenderSpec.hs:133]: > _At 01:00, 12 March 2023 in America/Havana, the clocks are turned forward 1 hour(s)_

This seems incorrect now :sweat_smile: 

The clocks are not turned forward at 01:00, they're turned forward at 00:00: https://www.timeanddate.com/time/change/cuba/havana?year=2023

File: test/Test/TzBot/RenderSpec.hs (hunk 4 above)
Comment (dcastro) [reply test/Test/TzBot/RenderSpec.hs]: Sweet, great work! :clap:

File: test/Test/TzBot/RenderSpec.hs (hunk 5 above)
Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:133]: Please let's just use plain strings because changing loads of well-formatted test entries is too painful.

File: test/Test/TzBot/RenderSpec.hs (hunk 4 above)
Comment (dcastro) [reply test/Test/TzBot/RenderSpec.hs]: Yes, exactly :D

File: test/Test/TzBot/RenderSpec.hs (hunk 5 above)
Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:133]: You are definitely right, but when I started doing it that way, strings were way not so long, and now it's already too late. I guess it would be better to fix in a separate PR (maybe along with sections that you wanted to delete, I still remember! :laughing:)

Comment (YuriRomanowski) [reply test/Test/TzBot/RenderSpec.hs:133]: Ouch

File: src/TzBot/TimeReference.hs [hunk 6]
```diff
@@ -143,24 +136,46 @@ timeReferenceToUTC sendersTZLabel eventTimestamp timeRef@TimeReference {..} =
     TZT.TZGap invalidTime _ _ -> TRTUInvalid $
       TimeShiftErrorInfo tzLabel (localDay invalidTime)
 
+  fromTZTimeWithFix :: Bool -> TZI.TZInfo -> LocalTime -> TZT.TZTime
+  fromTZTimeWithFix fixForward tzInfo localTime = case TZT.fromLocalTimeStrict tzInfo localTime of
+    Right res -> res
+    Left _ -> do
+      let diffHours = if fixForward then 3 else (-3)
+      let fix' = addLocalTime (TZT.hours diffHours)
```
Comment (dcastro) [inline src/TzBot/TimeReference.hs:144]: Why are we adding/subtracting 3 hours?

Comment (dcastro) [reply src/TzBot/TimeReference.hs:144]: It seems like simply using `TZT.fromLocalTime` on the "interval boundaries" (aka, the inferred date +/- 3 days) would be a good enough solution.

`TZT.fromLocalTime` applies a fair "adjustment" algorithm if the local time is invalid.

Comment (YuriRomanowski) [reply src/TzBot/TimeReference.hs:144]: If we get `TZError`, then it means we got into a timeshift, and we want to collect it for sure to emit a warning, that's why boundaries are adjusted this way (lower becomes lower, higher becomes higher)

Comment (YuriRomanowski) [reply src/TzBot/TimeReference.hs:144]: Well, sometimes `fromLocalTime` can leave a timeshift outside of the interval, that's why I made my own version.

Comment (dcastro) [reply src/TzBot/TimeReference.hs]: > Well, maybe that's not so important for us

Yea, I think so.

I mean, we want to perform the check in a `[-3 days, +3 days]` interval (aka, a 6 days interval).

But this is a completely arbitrary interval, there's nothing special about the number 6. We could just as easily have picked an 8 day interval.

So, if the boundaries of that interval happen to land exactly on top of a time shift, and `fromLocalTime` moves them a bit such that we now have a `[-2 days and 23hours, +3 days]` interval (aka, a 5 days and 23 hours interval), or a `[-3 days, +3 days and 1 hour]`, it's not a big deal, the margin is still quite wide.

File: src/TzBot/TimeReference.hs [hunk 7]
```diff
@@ -144,23 +130,37 @@ timeReferenceToUTC sendersTZLabel eventTimestamp timeRef@TimeReference {..} =
...
+            if thatTimeOfCurrentDay >= eventLocalTime
+            then eventLocalTime
+            else TZT.addCalendarClip (TZT.calendarDays 1) eventLocalTime
+          checkDiff = 3
+          daysToCheckTimeshifts =
+            ( TZT.addCalendarClip (TZT.calendarDays (negate checkDiff)) res
+            , TZT.addCalendarClip (TZT.calendarDays checkDiff) res
+            )
+
+      (res, Just daysToCheckTimeshifts)
```
Comment (dcastro) [inline src/TzBot/TimeReference.hs:148]: This is getting really hard to read and review :/

Some examples of specific pain points are:

* `dayTransition` now mixes 2 different purposes: to calculate the date the user is referring to + to calculate the +/-3 days interval around that date. This should be split into two different functions, each with one well-defined purpose.
* Lack of documentation: for a reader who's new to the codebase, it's not easy to understand why we're calculating this interval in the first place.

I've gone ahead and wrote some docs for this feature [in this comment](https://github.com/serokell/tzbot/pull/63#discussion_r1116030513).

I also created a branch `diogo/warn-on-possible-timeshifts`, 041b86ba843c83368b0bf8a531b3a62556acbbf2, where I reverted `dayTransition` to its old definition, and extracted the "time shift" logic into a separate function `calculateTimeshiftCheckInterval`. Please have a look; if you like it, please copy those changes into this PR.

Comment (dcastro) [reply src/TzBot/TimeReference.hs:148]: Actually, I went one step further. I noticed the logic of this feature is split in 3 places:
* In `TimeReference.timeReferenceToUTC`, we decide which interval we're going to check for offset changes
* In `TZ.checkForTimeshifts`, we process this interval and actually check for offset changes
* In `Render.renderOnSuccess`, we call `TZ.checkForTimeshifts` and then render the result as a warning.

To avoid having this big split, I thought it would be better to unify the "decide which interval" and "process the interval" steps.

In other words, now `TZ.checkForTimeshifts` creates the interval and checks it for offset changes immediately.
`TimeReference.timeReferenceToUTC` does not care about this at all, it only cares about converting the time reference to UTC, and that's it.

I pushed another commit to my branch, 60c8020e02bcb3fbcb55d964cdd42f6c2c2e5afc, please have a look.

Comment (YuriRomanowski) [reply src/TzBot/TimeReference.hs]: Thank you very much! Looks good! (and also the ideal to which I should aspire :smile:)

**Review by dcastro — APPROVED** [review]
